src/main/kotlin/fr/xebia/xke/5_FP.kt

```diff
@@ -0,0 +1,75 @@
+package fr.xebia.xke
+
+import arrow.core.*
+import arrow.syntax.option.toEither
+
+import arrow.core.Either
+import arrow.core.Option
+import arrow.core.flatMap
+import arrow.data.NonEmptyList
+import arrow.data.Validated
+import arrow.data.applicative
+import arrow.data.ev
+import arrow.syntax.applicative.tupled
+import arrow.syntax.functor.map
+
+/**
+ * This class validates user data in two ways:
+ * - Applicative for all fields (validated all fields)
+ * - Monadic within each field (stops on the first error per field)
+ */
+class UserValidation {
+
+    /*
+     * The following functions might help you
+     * - Option.fromNullable("something")
+     * - Option(maybeRight).toEither { "Left value" }
+     * - Either.cond(true, { 42 }, { "Error" })
+     * - Validated.fromEither(...)
+     * - Validated.applicative<List<String>>()
+            .map2(validated1, validated2, { v: Tuple<Int, Int> -> v.a + v.b })
+            .ev()
+     * - NonEmptyList.of("a string", "another string")
+     */
+
+    // TODO find a way to aggregate in an applicative way the 'readName' and 'readAge' functions
+    fun validateUser(params: Map<String, String>): Validated<NonEmptyList<String>, User>  {
+        return Validated.applicative<NonEmptyList<String>>()
+            .tupled(
```

Comment:
I replaced `map2` by `tupled` to avoid confusion

Reply:
This is the method I was looking for in order to check 3 or more fields

Reply:
This is the method I was looking for in order to check 3 or more fields